### PR TITLE
WiP stats: New API for bootstrap to add a stats builtin

### PR DESCRIPTION
**Pull request by jmarantz**
Description: Configuration for stat names that are not predictable within the source code.
There are several scenarios where Envoy records stats based on names that
are not known at compile time. For example:
  - Exotic SSL ciphers
  - Dynamo tables
  - gRPC method names

In these cases, stat-names are discovered in the request path, and we want
to avoid taking a global symbol table lock. These names are discoverable
using admin endpoint `/stats/recentlookups`. If an administrator discovers
these names are being looked up frequently, the bootstrap file can be
updated to pre-allocate those, and then access them lock-free.

This should help de-risk #4980
Risk Level: medium
Testing: //test/...
Docs Changes: will be needed
Release Notes: will be needed

### CODE REVIEW

**Comment by jmarantz** [comment]
@htuch for API review and also a first pass on acceptability of a union-of-pointers mechanism to save memory.

File: source/common/stats/symbol_table_impl.h [hunk 1]
```diff
@@ -39,6 +39,152 @@ constexpr uint64_t StatNameMaxSize = 1 << (8 * StatNameSizeEncodingBytes); // 65
...
+ *
+ * This is intended for embedding in other data structures that have access
+ * to a SymbolTable. StatNameStorage::free(symbol_table) must be called prior
+ * to allowing the StatNameStorage object to be destructed, otherwise an assert
+ * will fire to guard against symbol-table leaks.
+ *
+ * Thus this class is inconvenient to directly use as temp storage for building
+ * a StatName from a string. Instead it should be used via StatNameManagedStorage.
+ */
+class StatNameStorage {
```
Comment (jmarantz) [inline source/common/stats/symbol_table_impl.h]: Note: StatNameStorage was re-ordered in the file, but was otherwise unchanged.

File: source/common/stats/symbol_table_impl.h [hunk 2]
```diff
@@ -39,6 +39,152 @@ constexpr uint64_t StatNameMaxSize = 1 << (8 * StatNameSizeEncodingBytes); // 65
...
+ * Example usage:
+ *   StatNamePool pool(symbol_table);
+ *   StatName name1 = pool.add("name1");
+ *   StatName name2 = pool.add("name2");
+ *   uint8_t* storage = pool.addReturningStorage("name3");
+ *   StatName name3(storage);
+ */
+class StatNamePool {
+public:
+  explicit StatNamePool(SymbolTable& symbol_table) : symbol_table_(symbol_table) {}
```
Comment (jmarantz) [inline source/common/stats/symbol_table_impl.h]: Note: StatNamePool was re-ordered in the file but otherwise unchanged.

File: source/common/stats/symbol_table_impl.h [hunk 3]
```diff
@@ -39,6 +39,152 @@ constexpr uint64_t StatNameMaxSize = 1 << (8 * StatNameSizeEncodingBytes); // 65
...
+  // We keep the stat names in a vector of StatNameStorage, storing the
+  // SymbolTable reference separately. This saves 8 bytes per StatName,
+  // at the cost of having a destructor that calls clear().
+  SymbolTable& symbol_table_;
+  std::vector<StatNameStorage> storage_vector_;
+};
+
+// Represents a map from string to StatName, including the pool used to allocate
+// new StatNames.
+class StatNameMappedPool {
```
Comment (jmarantz) [inline source/common/stats/symbol_table_impl.h]: Note: I owe unit tests for this class. Wanted to start getting feedback on the rest of the PR in parallel.

File: source/common/stats/symbol_table_impl.h [hunk 4]
```diff
@@ -417,81 +522,28 @@ StatName StatNameStorage::statName() const { return StatName(bytes_.get()); }
...
-  uint8_t* addReturningStorage(absl::string_view name);
+  StatName statName() const { return StatName(storage_.builtin_); }
 
 private:
-  // We keep the stat names in a vector of StatNameStorage, storing the
-  // SymbolTable reference separately. This saves 8 bytes per StatName,
-  // at the cost of having a destructor that calls clear().
-  SymbolTable& symbol_table_;
-  std::vector<StatNameStorage> storage_vector_;
+  union {
```
Comment (jmarantz) [inline source/common/stats/symbol_table_impl.h]: @htuch let me just point you directly to the most controversial part of this PR: this union of pointers.

I could also have avoided a union and used const_cast, but I felt like this expressed the semantics a little better.

File: source/common/stats/thread_local_store.h [hunk 5]
```diff
@@ -313,10 +313,6 @@ class ThreadLocalStoreImpl : Logger::Loggable<Logger::Id::stats>, public StoreRo
     absl::optional<std::reference_wrapper<const StatType>>
     findStatLockHeld(StatName name, StatMap<RefcountPtr<StatType>>& central_cache_map) const;
 
-    void extractTagsAndTruncate(StatName& name,
-                                std::unique_ptr<StatNameManagedStorage>& truncated_name_storage,
-                                std::vector<Tag>& tags, std::string& tag_extracted_name);
-
```
Comment (jmarantz) [inline source/common/stats/thread_local_store.h]: dead declaration removed. could put this in a different PR if you prefer.

File: api/envoy/config/metrics/v2/stats.proto [hunk 6]
```diff
@@ -40,6 +40,24 @@ message StatsSink {
...
+//
+//   - Exotic SSL ciphers
+//   - Dynamo tables
+//   - gRPC method names
+//
+// In these cases, stat-names are discovered in the request path, and we want
+// to avoid taking a global symbol table lock. These names are discoverable
+// using admin endpoint `/stats/recentlookups`. If an administrator discovers
+// these names are being looked up frequently, the bootstrap file can be
+// updated to pre-allocate those, and then access them lock-free.
```
Comment (htuch) [inline api/envoy/config/metrics/v2/stats.proto]: Is this best effort? I imagine some things like cluster might be very dynamic, e.g. generated by the control plane, container environment, etc. and not be amenable to this.

File: source/common/stats/symbol_table_impl.h (hunk 4 above)
Comment (htuch) [reply source/common/stats/symbol_table_impl.h]: I think the complexity of stats is probably more controversial than this particular C++ styleism :) I'm very keen on seeing a stats fuzzer landed sooner rather than later for this reason. Here's the rationale:
1. While we have new and old versions, we can differentially fuzz and actually validate the new implementation is doing the right thing, continously as new optimizations are landed.
2. While stats might not be viewed as security critical, I actually think there are potential attacks via the control plane, e.g. if an untrusted party can name clusters or control the sequence of add/removal.

I think such a fuzzer can be written in O(1-2 hours). If we landed this, it would make it easier for review (cough).

**Comment by mattklein123** [comment]
Stepping back, would it be better to put this feature on hold until we know for sure we have a perf problem? As @htuch mentioned this is all already ridiculously complicated and it's not clear to me if we should add more complexity unless there is a proven need? Thoughts?

**Comment by jmarantz** [comment]
Sounds good; going to mark WiP for now and will (a) do some benchmarking and (b) do some fuzzing.

**Comment by jmarantz** [comment]
Closing this PR for now in favor of https://github.com/envoyproxy/envoy/pull/9355 which eliminates locks entirely for dynamically discovered stats.
